File: scripts/check.py
#!/usr/bin/env python3
# check.py — Pre-flight manifest validation
# Verifies that the manifest is valid and all clusters can be loaded
# without actually computing fingerprints.
#
# Usage:
#   python scripts/check.py
#   python scripts/check.py --cluster my-cluster

import sys
import os
import json
import importlib

def parse_args():
    args = sys.argv[1:]
    cluster_filter = None
    i = 0
    while i < len(args):
        if args[i] == '--cluster' and i + 1 < len(args):
            cluster_filter = args[i + 1]
            i += 2
        else:
            i += 1
    return cluster_filter


def main():
    cluster_filter = parse_args()
    manifest_path = os.path.join(os.getcwd(), 'regrets', 'manifest.json')
    
    # ─── Check manifest file ─────────────────────────────────────────────────
    if not os.path.exists(manifest_path):
        print('❌ regrets/manifest.json not found')
        print('   Run `regret init` to create it')
        sys.exit(1)
    
    try:
        with open(manifest_path, 'r', encoding='utf-8') as f:
            manifest = json.load(f)
    except json.JSONDecodeError as e:
        print(f'❌ manifest.json is invalid JSON: {e}')
        sys.exit(1)
    
    if 'clusters' not in manifest:
        print('❌ manifest.json missing "clusters" key')
        sys.exit(1)
    
    clusters = manifest['clusters']
    if cluster_filter:
        clusters = [c for c in clusters if c['id'] == cluster_filter]
    
    if not clusters:
        print(f'❌ No clusters found{f" matching {cluster_filter}" if cluster_filter else ""}')
        sys.exit(1)
    
    print(f'\n🔍 Checking {len(clusters)} cluster(s)...\n')
    
    # ─── Add pythonPath entries ──────────────────────────────────────────────
    python_paths = set()
    for cluster in clusters:
        pp = cluster.get('pythonPath', '')
        if pp:
            abs_path = os.path.join(os.getcwd(), pp)
            if abs_path not in sys.path:
                sys.path.insert(0, abs_path)
                python_paths.add(abs_path)
    
    # Support multiple pythonPaths (comma-separated or list)
    # This is a new feature: pythonPath can be a list of directories
    for cluster in clusters:
        pp = cluster.get('pythonPaths', [])
        if isinstance(pp, str):
            pp = [p.strip() for p in pp.split(',')]
        for p in pp:
            abs_path = os.path.join(os.getcwd(), p)
            if abs_path not in sys.path:
                sys.path.insert(0, abs_path)
                python_paths.add(abs_path)
    
    passed = 0
    failed = 0
    warnings = 0
    
    for cluster in clusters:
        cid = cluster.get('id', '')
        entry = cluster.get('entry', '')
        module_path = cluster.get('module', cluster.get('file', ''))
        stack = cluster.get('stack', 'js')
        watches = cluster.get('watches', [])
        
        print(f'  Checking: {cid}')
        errors = []
        warns = []
        
        # Check required fields
        if not cid:
            errors.append('missing "id"')
        if not entry:
            errors.append('missing "entry"')
        if not module_path:
            errors.append('missing "module" or "file"')
        
        # Check Python stack specifics
        if stack == 'python':
            # Check module can be imported
            try:
                mod = importlib.import_module(module_path)
            except ImportError as e:
                errors.append(f'module "{module_path}" import failed: {e}')
            except Exception as e:
                errors.append(f'module "{module_path}" load error: {e}')
            else:
                class_method = cluster.get('classMethod')
                constructor_name = cluster.get('constructor', entry)
                setup_steps = cluster.get('setup', [])
                constructor_args = cluster.get('constructorArgs', [])

                if class_method:
                    # ── classMethod validation ─────────────────────────────
                    Cls = getattr(mod, constructor_name, None)
                    if Cls is None:
                        errors.append(f'constructor "{constructor_name}" not found in {module_path}')
                    elif not isinstance(Cls, type):
                        errors.append(f'constructor "{constructor_name}" is not a class (type: {type(Cls).__name__})')
                    else:
                        # Check the target method exists on the class
                        # (instantiate first to get bound methods)
                        try:
                            instance = Cls(*constructor_args)
                            target = getattr(instance, class_method, None)
                            if target is None or not callable(target):
                                errors.append(f'classMethod "{class_method}" not found on {constructor_name} instance')
                            else:
                                warns.append(f'classMethod OK: {constructor_name}.{class_method}()')
                        except Exception as e:
                            errors.append(f'constructor {constructor_name}() failed: {e}')

                        # Validate setup steps
                        for i_step, step in enumerate(setup_steps):
                            step_method_name = step.get('method', '')
                            step_args = step.get('args', [])

                            try:
                                instance = Cls(*constructor_args)
                            except Exception as e:
                                errors.append(f'constructor {constructor_name}() failed during setup check: {e}')
                                continue

                            setup_fn = getattr(instance, step_method_name, None)
                            if setup_fn is None or not callable(setup_fn):
                                errors.append(f'setup[{i_step}].method "{step_method_name}" not found on {constructor_name}')
                                continue

                            # Validate args format: if the method only takes
                            # keyword args (like classical(**kwargs)), args must be a dict
                            try:
                                import inspect
                                sig = inspect.signature(setup_fn)
                                params = list(sig.parameters.values())
                                
                                # Skip 'self' parameter
                                params = [p for p in params if p.name != 'self']
                                
                                has_positional = any(
                                    p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD)
                                    and p.default is p.empty
                                    for p in params
                                )
                                has_keyword_only = any(p.kind == p.KEYWORD_ONLY for p in params)
                                
                                if isinstance(step_args, list) and len(step_args) > 0:
                                    if has_keyword_only and not has_positional:
                                        errors.append(
                                            f'setup[{i_step}].args is a list but '
                                            f'{step_method_name}() only accepts keyword args — '
                                            f'change args to a dict: {{"key": value}}'
                                        )
                                    else:
                                        # Try calling with the args to validate
                                        try:
                                            setup_fn(*step_args)
                                        except TypeError as te:
                                            errors.append(
                                                f'setup[{i_step}].args failed on {step_method_name}(): {te}'
                                            )
                                elif isinstance(step_args, dict):
                                    # Try calling with kwargs
                                    try:
                                        setup_fn(**step_args)
                                    except TypeError as te:
                                        errors.append(
                                            f'setup[{i_step}].args failed on {step_method_name}(**{step_args}): {te}'
                                        )
                            except Exception as e:
                                warns.append(f'setup[{i_step}] validation error: {e}')

                        # Check watches exist on the class (not module level)
                        for watch in watches:
                            watch_on_class = getattr(Cls, watch, None)
                            watch_on_instance = None
                            try:
                                inst = Cls(*constructor_args)
                                watch_on_instance = getattr(inst, watch, None)
                            except Exception:
                                pass
                            if watch_on_class is None and watch_on_instance is None:
                                warns.append(f'watch "{watch}" not found on {constructor_name} class or instance')
                            elif watch_on_instance is not None and not callable(watch_on_instance):
                                warns.append(f'watch "{watch}" on {constructor_name} is not callable')

                else:
                    # ── Function-based entry validation ───────────────────
                    entry_fn = getattr(mod, entry, None)
                    if entry_fn is None or not callable(entry_fn):
                        entry_cls = getattr(mod, entry, None)
                        if entry_cls is not None and isinstance(entry_cls, type):
                            warns.append(f'entry "{entry}" is a class, not a function — consider using classMethod')
                        else:
                            errors.append(f'entry "{entry}" not found or not callable in {module_path}')
                    
                    # Check watches exist in module
                    for watch in watches:
                        watch_attr = getattr(mod, watch, None)
                        if watch_attr is None:
                            warns.append(f'watch "{watch}" not found in {module_path}')
                        elif not callable(watch_attr) and not isinstance(watch_attr, type):
                            warns.append(f'watch "{watch}" is not callable (type: {type(watch_attr).__name__})')
        
        # Check outputTransform validity
        output_transform = cluster.get('outputTransform')
        if output_transform and isinstance(output_transform, str):
            valid_transforms = ['get_val_d', 'to_dict', 'to_bytes', 'repr', 'hex']
            if output_transform not in valid_transforms:
                warns.append(f'outputTransform "{output_transform}" is not a known named transform (valid: {valid_transforms})')
        
        # Report results
        if errors:
            print(f'    ❌ FAIL')
            for e in errors:
                print(f'       {e}')
            for w in warns:
                print(f'       ⚠️  {w}')
            failed += 1
        elif warns:
            print(f'    ⚠️  WARN')
            for w in warns:
                print(f'       {w}')
            warnings += 1
            passed += 1
        else:
            print(f'    ✅ OK')
            passed += 1
    
    # ─── Summary ─────────────────────────────────────────────────────────────
    print(f'\n{"─" * 50}')
    print(f'Check complete: {passed} OK, {warnings} warnings, {failed} failed')
    
    if failed > 0:
        print(f'\n⚠️  Fix errors before running capture.')
        sys.exit(1)
    
    print(f'\n✅ All clusters pass pre-flight check. Safe to run capture.')
    sys.exit(0)


if __name__ == '__main__':
    main()

File: scripts/check.js
#!/usr/bin/env node
// check.js — Pre-flight manifest validation
// Verifies manifest structure AND that all entry functions exist
// in the compiled output before running capture, preventing confusing errors.
//
// Validation phases:
//   Phase 1 — Manifest structure validation (schema-level checks)
//   Phase 2 — Export existence validation (import & verify)
//
// Usage:
//   node scripts/check.js
//   node scripts/check.js --cluster my-cluster

import { readFileSync } from 'fs'
import { resolve } from 'path'
import { pathToFileURL } from 'url'
import { mergeCjsModule } from './cjs-merge.js'

const args = process.argv.slice(2)
const clusterFilter = args.includes('--cluster') ? args[args.indexOf('--cluster') + 1] : null
const manifestPath = args.includes('--manifest')
  ? args[args.indexOf('--manifest') + 1]
  : resolve(process.cwd(), 'regrets/manifest.json')

let manifest
try {
  manifest = JSON.parse(readFileSync(manifestPath, 'utf8'))
} catch {
  console.error(`❌ Could not read manifest: ${manifestPath}`)
  process.exit(1)
}

const clusters = clusterFilter
  ? manifest.clusters.filter(c => c.id === clusterFilter)
  : manifest.clusters

if (!clusters.length) {
  console.error(`❌ No clusters found${clusterFilter ? ` matching "${clusterFilter}"` : ''}`)
  process.exit(1)
}

console.log(`\n🔍 Pre-flight Check — ${clusters.length} cluster(s)\n`)

// ─── Phase 1: Manifest structure validation ──────────────────────────────────
// Collect ALL errors and warnings first, then print them at once.

const VALID_STACKS = ['js', 'ts', 'python', 'react', 'vue', 'go', 'php', 'rust', 'css']
const VALID_FINGERPRINT_LEVELS = ['entry', 'full', 'watched']

const structErrors = []   // { clusterId, message }
const structWarnings = [] // { clusterId, message }

// 1. Duplicate cluster IDs
const idCounts = new Map()
for (const cluster of clusters) {
  const id = cluster.id
  if (!id) continue
  idCounts.set(id, (idCounts.get(id) || 0) + 1)
}
for (const [id, count] of idCounts) {
  if (count > 1) {
    structErrors.push({ clusterId: id, message: `Duplicate cluster id: '${id}'. IDs must be unique.` })
  }
}

// Per-cluster validations
for (let i = 0; i < clusters.length; i++) {
  const cluster = clusters[i]
  const cid = cluster.id || `index-${i}`

  // 5. Missing required fields: id, entry, stack
  for (const field of ['id', 'entry', 'stack']) {
    if (!cluster[field]) {
      structErrors.push({ clusterId: cid, message: `cluster at index ${i} missing required field: ${field}` })
    }
  }

  // 4. Invalid stack value
  if (cluster.stack && !VALID_STACKS.includes(cluster.stack)) {
    structErrors.push({ clusterId: cid, message: `Unknown stack '${cluster.stack}'. Valid: ${VALID_STACKS.join('|')}` })
  }

  // 6. Invalid fingerprintLevel
  if (cluster.fingerprintLevel && !VALID_FINGERPRINT_LEVELS.includes(cluster.fingerprintLevel)) {
    structErrors.push({ clusterId: cid, message: `invalid fingerprintLevel '${cluster.fingerprintLevel}'` })
  }

  // 2. Empty inputs array
  if (cluster.inputs && Array.isArray(cluster.inputs) && cluster.inputs.length === 0) {
    structWarnings.push({ clusterId: cid, message: `cluster '${cid}' has no inputs — fingerprint will be empty` })
  }
  if (!cluster.inputs && !cluster.multiArgs) {
    // No inputs field at all — also warn
    structWarnings.push({ clusterId: cid, message: `cluster '${cid}' has no inputs — fingerprint will be empty` })
  }

  // 3. Empty watches array with fingerprintLevel:'full'
  if (cluster.fingerprintLevel === 'full') {
    const watches = cluster.watches
    if (!watches || !Array.isArray(watches) || watches.length === 0) {
      structErrors.push({ clusterId: cid, message: `watches required for fingerprintLevel:'full'` })
    }
  }

  // 7. multiArgs:true but inputs not array of arrays
  if (cluster.multiArgs) {
    const inputs = cluster.inputs
    if (!Array.isArray(inputs) || inputs.length === 0 || !inputs.every(item => Array.isArray(item))) {
      structWarnings.push({ clusterId: cid, message: `multiArgs requires inputs to be array of arrays` })
    }
  }
}

// Print Phase 1 results
if (structErrors.length > 0 || structWarnings.length > 0) {
  console.log('📋 Phase 1 — Manifest structure validation:\n')
  for (const err of structErrors) {
    console.error(`  Error: [${err.clusterId}] ${err.message}`)
  }
  for (const warn of structWarnings) {
    console.warn(`  Warning: [${warn.clusterId}] ${warn.message}`)
  }
  console.log()
}

// If structural errors exist, stop here (exit 1)
if (structErrors.length > 0) {
  console.log(`${'─'.repeat(50)}`)
  console.log(`Manifest structure: ${structErrors.length} error(s), ${structWarnings.length} warning(s)`)
  console.log(`\n❌ Fix manifest errors before running capture.`)
  process.exit(1)
}

// If only warnings, note them but continue
if (structWarnings.length > 0) {
  console.log(`  (${structWarnings.length} warning(s) — non-blocking)\n`)
}

// If no errors and no warnings, print success for phase 1
if (structErrors.length === 0 && structWarnings.length === 0) {
  console.log(`📋 Phase 1 — Manifest structure: ✅ ${clusters.length} cluster(s), all fields OK\n`)
}

// ─── Phase 2: Export existence validation ────────────────────────────────────

// Run preBuild if specified
if (manifest.preBuild) {
  console.log(`🔧 Running preBuild: ${manifest.preBuild}`)
  const { execSync } = await import('child_process')
  try {
    execSync(manifest.preBuild, { stdio: 'inherit', cwd: process.cwd() })
  } catch {
    console.error(`❌ preBuild failed`)
    process.exit(1)
  }
}

let passed = 0
let warnings = structWarnings.length
let failed = 0

// Group by file to avoid redundant imports
const fileMap = new Map()
for (const cluster of clusters) {
  const { file, stack } = cluster
  if (stack && stack !== 'js' && stack !== 'ts') {
    console.log(`  ⏭️  ${cluster.id}: stack="${stack}" — skip (use native validator)`)
    continue
  }
  if (!fileMap.has(file)) fileMap.set(file, [])
  fileMap.get(file).push(cluster)
}

for (const [file, fileClusters] of fileMap) {
  console.log(`\n📂 ${file}`)

  let mod
  try {
    const absPath = resolve(process.cwd(), file)
    const moduleUrl = pathToFileURL(absPath).href
    let rawModule = await import(moduleUrl)
    mod = mergeCjsModule(rawModule)
  } catch (e) {
    console.error(`  ❌ Cannot import module: ${e.message}`)
    for (const c of fileClusters) {
      console.error(`     ${c.id}: FAIL (module import failed)`)
      failed++
    }
    continue
  }

  const availableExports = Object.keys(mod).filter(k => typeof mod[k] === 'function')
  console.log(`   Available functions: ${availableExports.length}`)

  for (const cluster of fileClusters) {
    const { id, entry, watches = [], classMethod, constructor: constructorName, storeDispatch } = cluster

    // Check storeDispatch mode
    if (storeDispatch) {
      const storeExport = mod[storeDispatch.store] ?? mod.default?.[storeDispatch.store]
      if (!storeExport) {
        console.error(`  ❌ ${id}: Store "${storeDispatch.store}" not found`)
        failed++
        continue
      }
      if (typeof storeExport.dispatch !== 'function') {
        console.error(`  ❌ ${id}: "${storeDispatch.store}" has no dispatch method`)
        failed++
        continue
      }
      console.log(`  ✅ ${id}: store "${storeDispatch.store}" found with dispatch`)
      passed++
      continue
    }

    // Check classMethod mode
    if (classMethod) {
      const Cls = mod[constructorName ?? entry] ?? mod.default?.[constructorName ?? entry]
      if (typeof Cls !== 'function') {
        console.error(`  ❌ ${id}: Constructor "${constructorName ?? entry}" not found or not a class`)
        failed++
        continue
      }
      console.log(`  ✅ ${id}: class "${constructorName ?? entry}" found`)
      passed++
      continue
    }

    // Check function-based entry
    const entryFn = mod[entry]
      ?? mod.default?.[entry]
      ?? ((entry === 'default' || entry === 'module.exports') && typeof mod.default === 'function' ? mod.default : null)

    if (typeof entryFn !== 'function') {
      // Suggest similar names
      const suggestions = availableExports.filter(k =>
        k.toLowerCase().includes(entry.toLowerCase().slice(0, 5)) ||
        entry.toLowerCase().includes(k.toLowerCase().slice(0, 5))
      )
      const suggestStr = suggestions.length
        ? ` (did you mean: ${suggestions.join(', ')}?)`
        : ''
      console.error(`  ❌ ${id}: Entry "${entry}" not found${suggestStr}`)
      failed++
      continue
    }

    // Check watches
    const missingWatches = watches.filter(w => typeof mod[w] !== 'function')
    if (missingWatches.length > 0) {
      console.warn(`  ⚠️  ${id}: Watch target(s) not found as functions: ${missingWatches.join(', ')}`)
      warnings++
    }

    console.log(`  ✅ ${id}: entry "${entry}" found, ${watches.length} watch(es) checked`)
    passed++
  }
}

console.log(`\n${'─'.repeat(50)}`)
console.log(`Pre-flight: ${passed} passed, ${warnings} warnings, ${failed} failed`)

if (failed > 0) {
  console.log(`\n❌ Fix failed checks before running capture.`)
  process.exit(1)
}

console.log(`\n✅ Manifest valid: ${clusters.length} clusters, all fields OK`)
